### ISSUE TITLE
RFC: Keyboard Shortcuts API

### ISSUE THREAD

**Issue by bih** [spotify/backstage#431]
_This is mostly based on empirical knowledge (and some inspiration from Visual Studio Code) rather than any scientific or user experience evidence, so please do chime in. I would love to hear from you. This will definitely iterate over many rounds of comments and will likely change significantly._

Being able to navigate across Backstage with keyboard shortcuts is more productive for regular users, as well as providing additional interfaces for impaired users. I have dabbled quite a bit with this in Spotify's internal use of Backstage, although it was never finished nor ever made it to production.

## Concept

In this proposal, there will be two types of shortcuts: App Shortcuts and Plugin Shortcuts. Both will need to register a shortcut by sharing an unique ID, a function callback, and a preferred shortcut key. More importantly, registering a shortcut **does not mean** the shortcut will be enabled. Users will need to manually enable any shortcuts in Backstage for two reasons: a) for user education and b) to enable them to customize shortcuts from the very start to suit their flow (i.e. Vim, etc)

### App Shortcuts

**App Shortcuts** are registered by the App (`packages/app` folder). This is usually those who deploys Backstage for their team/company. They will have the ability to register shortcuts for across Backstage, and will be for common tasks such as navigating to top-level views, creating new components, etc.

### Plugin Shortcuts

**Plugin Shortcuts** are registered by any Plugin (`plugins/*` folders). They will be able to register a limited number of shortcuts (let's say 2 for now) in their `plugin.ts` manifest. The shortcut ID will be scoped against their plugin in someway (maybe: prepend the shortcut ID with plugin ID). The shortcuts will only be available within the context of their plugin.

### Shortcuts UI

As mentioned, both App and Plugin Shortcuts will only register shortcuts. No shortcuts will be enabled by default with this API. Users can navigate to a Shortcuts UI where they can view all available keyboard shortcuts, their preferred shortcut (along with other possible metadata such as descriptions), and the ability one-click enable/customize them. Plugin Shortcuts will only work within the context of their plugin. The user's settings will be in persisted storage.

## Possible Implementation

Here's some psuedocode for inspiration. Although mostly up to your imagination :)

```ts
const shortcuts = new ShortcutsBuilder();
shortcuts.registerGlobal('docs', () => history.push('/docs'),  { preferred: 'cmd+d' });
builder.add(shortcuts.build());
```

```ts
const app = createApp({
  enabledShortcuts: ['welcome.reload-page']
})
```

```tsx
import { createPlugin } from '@backstage/core';
import WelcomePage from './components/WelcomePage';

export default createPlugin({
  id: 'welcome',
  register({ router, keyboardShortcuts }) {
    router.registerRoute('/', WelcomePage);

    keyboardShortcuts.register('welcome.reload-page', { preferred: 'cmd+d' });
  },
});
```

```tsx
import React from 'react';
import { useKeyboardShortcutPress } from '@backstage/core';

export const ExampleComponent: React.FC<{}> = () => {
  useKeyboardShortcutPress(({ shortcutId, combination }) => {
    // do something here
    if (shortcutId === 'welcome.reload-page') {
      window.location.reload();
    }
  });

  // or

  useKeyboardShortcutPress('welcome.reload-page', () => {
    window.location.reload()
  });

  return <h1>Hello world</h1>;
});
```

## Possible Considerations

- <s>**Double key presses.** How could we handle double key presses, or do we want to at all?</s> Edit: Removed from the RFC.
- <s>**App Shortcuts Defaults.** Do we want to have an out-of-the-box set of keyboard shortcuts (i.e. <kbd>Cmd+D</kbd>) to go to the Docs view?</s> Edit: Added to Possible Implementation.
- **Limit which keys/combinations can be enabled.** From my experience building out a prototype of this, there were some keys that were a no-no. Maybe it would be great if Backstage could have an "index" of key combinations that plugins can use to avoid any common/obvious keyboard shortcut pitfalls?
- **Limit how many shortcuts plugins can register.** This would be good to keep the number of shortcuts at a minimum at first, and over time, being able to scale this out as the UI develops. The purpose is not to shortcut everything, but only a few hand-selected tasks.
- **Ability to share configurations.** Do we want to enable exporting of a user's keybindings (maybe in a JSON format?), it might be useful in an organizational-wide use case (unlikely to be useful in the greater open source community). An example might be having a Vim-like bindings configuration within Spotify's Backstage instance.

**Comment by bih** [comment]
@backstage/core any comments/feedback?

**Comment by bih** [comment]
@spotify/backstage-core any comments?

**Comment by Rugvip** [comment]
Thanks for the RFC!

Gonna preface this with that this is going to be pretty low prio as we don't really see a need to get this shipped early on and there are many other APIs that we'd prioritise. That said it's still good to get the discussion going :grin:

I'm interested in how `onKeyboardShortcutPress` would work, is it a react-hook? Should be something similar to the feature flags API.

### Considerations

> * Double key presses.

Don't think we need to support this tbh, especially not from the start.

> * Limit which keys/combinations can be enabled.

Yup, and also be deliberate with when we capture input. Single-key hotkeys like `s` would be fine if we don't capture in input fields, but we'd need to add modifiers if we wanna capture in input fields, and `ctrl+s` would be a bad one :grin:.

> * Limit how many shortcuts plugins can register.

Don't really see why this is needed? A guy would probably group the shortcuts per plugin anyway, and they have to be explicitly enabled?

> * App Shortcuts Defaults.

I think we could allow the app to enable some shortcuts by default. Smth like

```ts
const app = createApp({
  enabledShortcuts: ['welcome.reload-page']
})
```

> * Ability to share configurations.

Probably rely on the app enabling things for orgs instead tbh

**Comment by bih** [comment]
> Gonna preface this with that this is going to be pretty low prio as we don't really see a need to get this shipped early on and there are many other APIs that we'd prioritise. That said it's still good to get the discussion going 😁

Makes sense! Is there any particular tag this should be to indicate its priority? If so, feel free to tag. The reason I decided to create an RFC is because @stefanalund asked me directly about it. I don't plan on working on this anytime soon but would leave it here if anyone else wanted to pick up / discuss 😄
 
> I'm interested in how `onKeyboardShortcutPress` would work, is it a react-hook? Should be something similar to the feature flags API.

Technically yes, although under the hood it can be non-React code. I see the hook as a wrapper, so that the underlying code can be ported to any other UI framework or away from Backstage entirely.
 
> ### Considerations
> > * Double key presses.
> 
> Don't think we need to support this tbh, especially not from the start.

Good to hear. I'll remove this.

> > * Limit which keys/combinations can be enabled.
> 
> Yup, and also be deliberate with when we capture input. Single-key hotkeys like `s` would be fine if we don't capture in input fields, but we'd need to add modifiers if we wanna capture in input fields, and `ctrl+s` would be a bad one 😁.

Yep makes sense. I think if there's any good guides or list of keys we'd want that we could (or anyone else reading) could contribute, that'd be great.
 
> > * Limit how many shortcuts plugins can register.
>
> Don't really see why this is needed? A guy would probably group the shortcuts per plugin anyway, and they have to be explicitly enabled?

In my view it was less about the technical limitation and more about the UI one. If every plugin had 10 shortcuts enabled, and there were 20 plugins say, that's 200 shortcuts. If this is something this feature wants to promote from day 0, then this point is not applicable. The fact that users have to manually opt into them might be sufficient enough.

> > * App Shortcuts Defaults.
> 
> I think we could allow the app to enable some shortcuts by default. Smth like
> 
> ```ts
> const app = createApp({
>   enabledShortcuts: ['welcome.reload-page']
> })
> ```

That's good feedback 👍 

> > * Ability to share configurations.
> 
> Probably rely on the app enabling things for orgs instead tbh

Not sure I understand this particular point. The app enabling things for orgs?

**Comment by Rugvip** [comment]
> Technically yes, although under the hood it can be non-React code. I see the hook as a wrapper, so that the underlying code can be ported to any other UI framework or away from Backstage entirely.

If it's a custom hook, it should be prefixed with `use`, or everyone's gonna have a bad time ⛷ 

I'd expect a simple hook implementation to look something like this:

```ts
function useKeyboardShortcut(shortcutId: string, callback: () => void) {
  const api = useApi(keyboardShortcutApi)

  useEffect(() => {
    const unsubscribe = api.subscribe(shortcutId, callback)
    return () => unsubscribe()
  }, [shortcutId, callback, api])
}
```

> Not sure I understand this particular point. The app enabling things for orgs?

I meant the feature from the previous point, orgs enabling shortcuts by default in the app.

**Comment by bih** [comment]
Ah yes, my bad - totally slipped that I used `on` instead of `use` :P good catch. In terms of the org enabling shortcuts, that makes sense for common/app shortcuts. I think the sharing configurations was more for the plugin shortcuts, but that might be adding premature complexity so probably not worth adding in the first iteration of this.

**Comment by ericis** [comment]
@bih if there are future plug-ins for embedded [VS Code](http://code.visualstudio.com/), [Theia](https://theia-ide.org/) or other editor, these shortcuts might then cause conflicts.

Perhaps it would be better to leave shortcuts out of Backstage and instead wait for plug-ins to evolve further to better understand the options? Of course, waiting could also mean never being able to implement shortcuts, because of breaking backwards compatibility with plug-ins that have developed their own shortcuts.

So, maybe the solution is to define very simple shortcuts for frequent interactions, like menu selections, etc.

**Comment by bih** [comment]
> At some point we had mockups for pinned items in the sidebar, that were essentially personal numbered shortcuts into the app. Those would be a good fit for app level mapped e.g. Ctrl+Shift+number or so.

That’s awesome! Yeah definitely can be a part of this

**Comment by meledin** [comment]
Keyboard accessibility in web applications. So fun. A few items that may be of interest.

Keyboard shortcuts in webapps, typically, are few and far between mainly due to conflicts with the browser's short keys. There is actually an [accesskey](https://html.spec.whatwg.org/#the-accesskey-attribute) attribute specifically for shortcuts, but the general consensus is that it _should not be used_ (because of a messy implementation).

What you'll find used quite a lot (you can see it here on GitHub even) is tab, shift-tab, arrow keys, space, enter. Also fairly common in developer tools is generic keypress with an input field not selected (e.g. pressing `c` while not in a text input field on GitHub leads to new issue creation).

Long story short, I would say you want to distinguish between _accessibility_ and _shortcuts_. Accessibility in a webapp is highly dependent on tab, tabindex, arrows, and space/enter. 

Shortcuts in a webapp are a hot mess and there's no currently working standard. So if global shortcuts are to be a thing, the very least you may want to do is reserve a namespace for them

**Comment by stefanalund** [comment]
Not prioritised.
